### ISSUE TITLE
Function casting not working on 1.11

### ISSUE THREAD

**Issue by geominorai** [alliedmodders/sourcepawn#671]
# Help us help you
  - [x] I have checked that my issue [doesn't exist yet](https://github.com/alliedmodders/sourcemod/issues).
  - [x] I have tried my absolute best to reduce the problem-space and have provided the absolute smallest test-case possible.
  - [x] I can always reproduce the issue with the provided description below.

# Environment
  * Current SourceMod snapshot: `1.11.0.6745`

# Description
Function casting previously working on the `1.10.0.6512` compiler no longer does on `1.11`.

# Problematic Code
```cpp
typedef MyFunc = function void ();

public int Native_Test(Handle hPlugin, int iArgC) {
	MyFunc fnFunc = view_as<MyFunc>(GetNativeFunction(1));
}
```

# Expected Behavior
```
SourcePawn Compiler 1.10.0.6502
Copyright (c) 1997-2006 ITB CompuPhase
Copyright (c) 2004-2018 AlliedModders LLC

C:\Build\SM1.10\scripting\spcomp_function_cast_bug.sp(4) : warning 204: symbol is assigned a value that is never used: "fnFunc"
Code size:             3056 bytes
Data size:             2280 bytes
Stack/heap size:      16384 bytes
Total requirements:   21720 bytes

1 Warning.
```

# Actual Behavior
```
SourcePawn Compiler 1.11.0.6745
Copyright (c) 1997-2006 ITB CompuPhase
Copyright (c) 2004-2018 AlliedModders LLC

C:\Build\SM1.10\scripting\spcomp_function_cast_bug.sp(4) : error 100: function prototypes do not match

1 Error.
```

**Comment by c0rp3n** [comment]
You should not be calling that function returned from `GetNativeFunction` as a normal function it should be going through `Call_StartFunction` if I remember correctly.

see:
https://sourcemod.dev/#/functions/function.Call_StartFunction

**Comment by geominorai** [comment]
> 
> 
> You should not be calling that function returned from `GetNativeFunction` as a normal function it should be going through `Call_StartFunction` if I remember correctly.
> 
> see:
> https://sourcemod.dev/#/functions/function.Call_StartFunction

Indeed, I can call the functions just fine via `Call_StartFunction` on servers running on SM 1.10 and 1.11.  It is just the 1.11 compiler misbehaving, so I am sticking with the 1.10 compiler for now until this gets fixed.

**Comment by asherkin** [comment]
> It is just the 1.11 compiler misbehaving, so I am sticking with the 1.10 compiler for now until this gets fixed.

I don't think it is, this was an intentional change:  #644

Why do you think that it is misbehaving? `GetNativeFunction` returns a `Function` and `Call_StartFunction` takes a `Function` - it doesn't appear to need to be a `MyFunc` at any point inside the native handler.

**Comment by geominorai** [comment]
I currently have two projects affected by this.

One project is a modular plugin set I am working on where the main plugin calls specific functions defined in a submodule's plugin, but the submodule first has to pass a pointer for one of its functions up to the main module during initialization via a native so it can be stored first.

The second is a project where I have an `enum struct` containing a function variable that defines a custom action, and that what that function points to is set up during initialization depending on the circumstances and later called.  This is done instead of having a huge `switch` block dispatching to different functions for each `case`.

**Comment by dvander** [comment]
The cast in the example is not needed, though.

**Comment by geominorai** [comment]
But what variable type can I store it as?  `function` is reserved and cannot be used as a type for variable declaration.

**Comment by dvander** [comment]
"Function" should work.

**Comment by geominorai** [comment]
Oh shit, I did not know that was a thing.  Thanks, I can definitely use this.

Is there a way to check or make sure the `Function` returned by `GetNativeFunction` match a specific signature similar to `typeset` / `typedef`?

**Comment by Wend4r** [comment]
> Is there a way to check or make sure the `Function` returned by `GetNativeFunction` match a specific signature similar to `typeset` / `typedef`?

No, the check takes place at the compilation level. Function signatures are not checked during run time.
You can represent `Function` by a type from `type(set/def)` via `view_as`.

**Comment by dvander** [comment]
I'm still open to making this cast work because we can add run-time checks for it later on.

The "Function" type vs "function" keyword is unfortunate. It's probably too late to revisit that though.

**Comment by geominorai** [comment]
> You can represent `Function` by a type from `type(set/def)` via `view_as`.

This stopped working on the 1.11 compiler too:

```cpp
typedef MyFunc = function void ();

public int Native_Test(Handle hPlugin, int iArgC) {
	Function fnFunc = GetNativeFunction(1);
	MyFunc fnMyFunc = view_as<MyFunc>(fnFunc);
}
```

```
SourcePawn Compiler 1.10.0.6512
Copyright (c) 1997-2006 ITB CompuPhase
Copyright (c) 2004-2018 AlliedModders LLC

C:\Build\SM1.10\scripting\spcomp_function_cast_bug.sp(5) : warning 204: symbol is assigned a value that is never used: "fnMyFunc"
Code size:             3084 bytes
Data size:             2280 bytes
Stack/heap size:      16384 bytes
Total requirements:   21748 bytes

1 Warning.
```

```
SourcePawn Compiler 1.11.0.6745                                                                         
Copyright (c) 1997-2006 ITB CompuPhase                                                                  
Copyright (c) 2004-2018 AlliedModders LLC                                                               
                                                                                                        
C:\Build\SM1.10\scripting\spcomp_function_cast_bug.sp(5) : error 100: function prototypes do not match  
                                                                                                        
1 Error.                                                                                                
```

**Comment by JoinedSenses** [comment]
Any updates on this? typedefs are unusable at the moment for situations where typed funcs are downgraded to `Function` type such as forwards and datapack storage. Workaround is to replace type with base `Function` type.

**Comment by geominorai** [comment]
I have resorted to passing Function around, ignoring any safety checks offered during compile-time, and then feeling very bad about it afterwards.

**Comment by Cruze03** [comment]
Any updates on this?
